module/zfs/dmu_send.c

```diff
@@ -587,15 +597,49 @@ dump_dnode(dmu_sendarg_t *dsp, const blkptr_t *bp, uint64_t object,
 		}
 	}
 
+	/*
+	 * The DRR_SPILL_BLOCK flag is set for every dnode which references
+	 * a spill block.  This allows the receiving pool to definitively
+	 * determine when a spill block should be kept or freed.
+	 */
+	if (dsp->dsa_spill && (dnp->dn_flags & DNODE_FLAG_SPILL_BLKPTR))
+		drro->drr_flags |= DRR_SPILL_BLOCK;
+
 	if (dump_record(dsp, DN_BONUS(dnp), bonuslen) != 0)
 		return (SET_ERROR(EINTR));
 
 	/* Free anything past the end of the file. */
 	if (dump_free(dsp, object, (dnp->dn_maxblkid + 1) *
 	    (dnp->dn_datablkszsec << SPA_MINBLOCKSHIFT), DMU_OBJECT_END) != 0)
 		return (SET_ERROR(EINTR));
+
+	/*
+	 * Send extra DRR_SPILL records.  By default, these are included in
+	 * the stream in order to recreate spill blocks that are incorrectly
+	 * removed.  This may occur in previous versions of ZFS when certain
+	 * attributes of the dnode change (block size, bonus size, etc).  This
+	 * scheme is fully backwards compatible.  The current implementation
+	 * of 'zfs recv' will correctly handle spill blocks and ignore these
+	 * these additional records.
+	 */
+	if (zfs_send_extra_spill_blocks &&
+	    (dnp->dn_flags & DNODE_FLAG_SPILL_BLKPTR) &&
+	    (DN_SPILL_BLKPTR(dnp)->blk_birth <= dsp->dsa_fromtxg)) {
```

Comment:
There are two main reasons I ended up adding it here:

1) The traversal logic correctly skips blocks born before the specified `td->td_min_txg`.  Rather than try and work an exception in to that code for these unmodified blocks with birth txgs below `td->td_min_txg` it seemed simpler to add it here.
2) We need to ensure the `DRR_SPILL` record occurs after the `DRR_OBJECT` record in the resulting stream.  Again, this made it easy.

If you can suggest a nice way to move this step in to the traversal logic I'm all for it.

Reply:
Yeah, on further inspection, doing it in dmu_traverse would be complex. The better alternative is probably to do it in send_cb (so it goes through the entire send pipeline), but we don't have any of the logic in redacted send integrated yet, which breaks down send_cb calls by type and creates records for each one. Under that model, it would be easy to just create two records when the object record happens.  I'll probably move this into send_cb when I rebase redacted send on top of this, in that case, so don't worry about this comment.